grudge/op.py

```diff
@@ -310,6 +326,109 @@ def mass(dcoll, *args):
     return _bound_mass(dcoll, dd)(u=vec)
 
 
+def reference_mass_matrix(actx, out_element_group, in_element_group):
+    @keyed_memoize_in(
+        actx, reference_mass_matrix,
+        lambda out_grp, in_grp: (out_grp.discretization_key(),
+                                 in_grp.discretization_key()))
+    def get_ref_mass_mat(out_grp, in_grp):
+        if out_grp == in_grp:
+            from meshmode.discretization.poly_element import mass_matrix
+
+            return actx.freeze(
+                actx.from_numpy(mass_matrix(out_grp))
+            )
+
+        from modepy import vandermonde
+        basis = out_grp.basis_obj()
+        vand = vandermonde(basis.functions, out_grp.unit_nodes)
+        o_vand = vandermonde(basis.functions, in_grp.unit_nodes)
+        vand_inv_t = np.linalg.inv(vand).T
+
+        weights = in_grp.quadrature_rule().weights
+        return actx.freeze(
+            actx.from_numpy(
+                np.einsum("j,ik,jk->ij", weights, vand_inv_t, o_vand)
+            )
+        )
+
+    return get_ref_mass_mat(out_element_group, in_element_group)
+
+
+def _apply_mass_operator(dcoll, dd_out, dd_in, vec):
+    from grudge.geometry import area_element
+
+    import numpy as np
+
+    in_discr = dcoll.discr_from_dd(dd_in)
+    out_discr = dcoll.discr_from_dd(dd_out)
+
+    actx = vec.array_context
+    area_elements = area_element(actx, dcoll, dd=dd_in)
+
+    data = []
+    # FIXME: I think there is a bug in loopy's implementation
+    # of einsum; this code works when dispatching to numpy. But
+    # fails when using actx.einsum due to strides mismatch
+    # on the mass matrix... (error raised when attempting to
+    # invoke the loopy kernel)
+    # NOTE: This ONLY fails when in_grp != out_grp
+    # TODO: Need to debug the loopy program...?
```

Comment:
@inducer I got the metric computations working! :tada:

But I think there is something wonky going on when using loopy's `einsum` implementation. I'm unable to pass the test in `test_new_world_grudge.py` (I get something like: `TypeError: strides mismatch on argument 'mass_mat' (after removing unit length dims...`).

It does, however, pass when I dispatch to `numpy.einsum` (as in the code down below shows). I rewrote the mass matrix routine to make it more debuggable.

I think I managed to track down where it's breaking; the exception is getting generated [here](https://github.com/inducer/loopy/blob/main/loopy/target/execution.py#L547), but I'm afraid it will take me some time to decipher this bit of loopy... I'll revisit this tomorrow.

Reply:
Unlike numpy, loopy-generated code is picky about the memory layout (strides) of the array it is given. (If you think about it, it *has* to be, given that it's generating C code!) You can manipulate this using [`tag_array_axes`](https://documen.tician.de/loopy/ref_transform.html?highlight=tag_array_axes#loopy.tag_array_axes), specifically. By default, loopy expects C-layout (row-major) arrays. Your matrix likely comes out of LAPACK and is therefore in Fortran (column-major) layout.

Reply:
@inducer Thanks for the info. I'll try to experiment with it.

BTW I forgot to mention that the test only fails when `in_grp != out_grp`. IOW: when the mass matrix is not square due to using a quadrature element group as the `in_grp`.